### PR TITLE
Minor code changes in `Model`

### PR DESCRIPTION
**Pull request by tristandeleu**
A few (minor) changes to the `Model` class, mostly to make use of already existing code and avoid duplicates.
- Move the evaluation on the validation set in the epoch loop (rather than in the batch loop for the last batch).
- Remove the batch index from the logs given to the `on_batch_begin()` and `on_batch_end()` callbacks -- duplicate from the `batch` input.
- Make the `history` callback explicit for clarity.
- Use the methods `Model.train()` and `Model.test()` in the `Model.fit()` and `Model.evaluate()` loops. As a consequence I separated explicitly the loss function and the accuracy function -- eg. instead of having `Model._train()` *or* `Model._train_with_acc()` there are `Model._train()` *and* `Model._train_accuracy()` to compute the accuracy separately.

This last point is actually a preparation for a next step: remove the accuracy computation from `Model`. I think it would be more flexible to have the metrics computation completely separated from the model, inspired by [what scikit-learn is doing](http://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics). And one could compute the metrics during training with callbacks. This would also allow the user to define her own metrics of interest (maybe she is not interested in the accuracy but in the precision per class, F1, BLEU, etc... motivated by https://github.com/fchollet/keras/issues/254#issuecomment-114418526).

Known issue with this PR:
- ~~The progress bar in `verbose=1` does not update for the last batch of every epoch (it may appear as computational *lag* but everything is fine)~~ (see tristandeleu/keras#1)

### CODE REVIEW

**Comment by fchollet** [comment]
> The progress bar in verbose=1 does not update for the last batch of every epoch (it may appear as computational lag but everything is fine)

This is the main reason why things are setup as they are currently (it makes perfect sense: you update the epoch callbacks right after the last batch of the epoch).

I think some of the things you propose are worth incorporating; please fix the above issue by reverting the changes.

**Comment by tristandeleu** [comment]
Sorry I was a bit short on time, that's why I only acknowledged this issue without solving it just yet. I thought it would make sense to have the validation at the epoch level and maybe make the progress bar work around it. I though of a really quick fix for this, that requires to explicitly mark the end of the line in the progress bar. See tristandeleu/keras#1. That's a little bit more verbose, so I can revert the change on the validation if you want to.

Also, what do you think of separating the metrics from `Model`? It may be valuable for the future but for it could raise some issues short term
- It would be impossible to log the accuracy (or any other metrics) with the base logger as callbacks do not interact with each other. One solution would be to let the user create her own logger on a different level of verbosity (possibly by inheriting from the current `BaseLogger`).
- We would have to think about how to access the validation set with callbacks.

**Comment by fchollet** [comment]
Let's revert for now. If you have a fix that you think is clearly a superior solution to what currently exists, you can submit it in a new PR and I'll review it.

**Comment by tristandeleu** [comment]
Ok I revert it :)

I will make a new PR for changes on the progress bar (along with this change on the evaluation of the validation set).

**Comment by tristandeleu** [comment]
Upon further tests it appears that these changes add a slight overhead. In particular on my low dimensional example with ~6000 batches, there is an extra 3s/epoch with this PR.

I close this PR and will investigate further.
